File: core/table.ts
import { Assertion } from "df/core/assertion";
import { ColumnDescriptors } from "df/core/column_descriptors";
import {
  Contextable,
  IActionConfig,
  IColumnsDescriptor,
  ICommonContext,
  IDependenciesConfig,
  IDocumentableConfig,
  INamedConfig,
  ITargetableConfig,
  Resolvable
} from "df/core/common";
import { Session } from "df/core/session";
import {
  checkExcessProperties,
  resolvableAsTarget,
  setNameAndTarget,
  strictKeysOf,
  toResolvable,
  validateQueryString
} from "df/core/utils";
import { dataform } from "df/protos/ts";

/**
 * @hidden
 */
export const TableType = ["table", "view", "incremental", "inline"] as const;
/**
 * Supported types of table actions.
 *
 * Tables of type `view` will be created as views.
 *
 * Tables of type `table` will be created as tables.
 *
 * Tables of type `incremental` must have a where clause provided. For more information, see the [incremental tables guide](guides/incremental-datasets).
 */
export type TableType = typeof TableType[number];

/**
 * @hidden
 */
export const DistStyleType = ["even", "key", "all"] as const;
/**
 * Valid types for setting the distribution style for Redshift tables.
 *
 * View the [Redshift documentation](https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_examples.html#r_CREATE_TABLE_NEW-diststyle-distkey-and-sortkey-options) for more information.
 */
export type DistStyleType = typeof DistStyleType[number];

/**
 * @hidden
 */
export const SortStyleType = ["compound", "interleaved"] as const;
/**
 * Valid types for setting the sort style for Redshift tables.
 *
 * View the [Redshift documentation](https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_examples.html#r_CREATE_TABLE_NEW-diststyle-distkey-and-sortkey-options) for more information.
 */
export type SortStyleType = typeof SortStyleType[number];

/**
 * Redshift-specific warehouse options.
 */
export interface IRedshiftOptions {
  /**
   * Sets the DISTKEY property when creating tables.
   *
   * For more information, read the [Redshift create table docs](https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_examples.html#r_CREATE_TABLE_NEW-diststyle-distkey-and-sortkey-options).
   */
  distKey?: string;

  /**
   * Set the DISTSTYLE property when creating tables.
   *
   * For more information, read the [Redshift create table docs](https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_examples.html#r_CREATE_TABLE_NEW-diststyle-distkey-and-sortkey-options).
   */
  distStyle?: string;

  /**
   * A list of string values that will configure the SORTKEY property when creating tables.
   *
   * For more information, read the [Redshift create table docs](https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_examples.html#r_CREATE_TABLE_NEW-diststyle-distkey-and-sortkey-options).
   */
  sortKeys?: string[];

  /**
   * Sets the style of the sort key when using sort keys.
   *
   * For more information, read the [Redshift sort style article](https://docs.aws.amazon.com/redshift/latest/dg/t_Sorting_data-compare-sort-styles.html).
   */
  sortStyle?: string;
}

const IRedshiftOptionsProperties = () =>
  strictKeysOf<IRedshiftOptions>()(["distKey", "distStyle", "sortKeys", "sortStyle"]);

/**
 * Snowflake-specific warehouse options.
 */
export interface ISnowflakeOptions {
  /**
   * If set to true, a secure view will be created.
   *
   * For more information, read the [Snowflake Secure Views docs](https://docs.snowflake.com/en/user-guide/views-secure.html).
   */
  secure?: boolean;

  /**
   * If set to true, a transient table will be created. Only applicable to actions of type "table".
   *
   * For more information, read the [Snowflake docs](https://docs.snowflake.com/en/user-guide/tables-temp-transient.html).
   */
  transient?: boolean;

  /**
   * A list of clustering keys to cluster the table by. Only applicable to actions of type "table" or "incremental".
   *
   * For more information, read the [Snowflake clustering docs](https://docs.snowflake.com/en/user-guide/tables-clustering-keys.html).
   */
  clusterBy?: string[];
}
const ISnowflakeOptionsProperties = () =>
  strictKeysOf<ISnowflakeOptions>()(["secure", "transient", "clusterBy"]);

/**
 * Azure SQL Data Warehouse-specific warehouse options.
 */
export interface ISQLDataWarehouseOptions {
  /**
   * The distribution option value.
   *
   * For more information, read the [Azure CTAS docs](https://docs.microsoft.com/en-gb/sql/t-sql/statements/create-table-as-select-azure-sql-data-warehouse?view=aps-pdw-2016-au7#examples-for-table-distribution).
   */
  distribution?: string;
}
const ISQLDataWarehouseOptionsProperties = () =>
  strictKeysOf<ISQLDataWarehouseOptions>()(["distribution"]);

/**
 * BigQuery-specific warehouse options.
 */
export interface IBigQueryOptions {
  /**
   * The key with which to partition the table. Typically the name of a timestamp or date column.
   *
   * For more information, read the [BigQuery partitioned tables docs](https://cloud.google.com/bigquery/docs/partitioned-tables).
   */
  partitionBy?: string;

  /**
   * The keys by which to cluster partitions by.
   *
   * For more information, read the [BigQuery clustered tables docs](https://cloud.google.com/bigquery/docs/clustered-tables).
   */
  clusterBy?: string[];

  /**
   * SQL based filter for when incremental updates are applied.
   *
   * For more information, see our [incremental dataset docs](https://docs.dataform.co/guides/incremental-datasets).
   */
  updatePartitionFilter?: string;

  /**
   * Key-value pairs for [BigQuery labels](https://cloud.google.com/bigquery/docs/labels-intro).
   *
   * If the label name contains special characters, e.g. hyphens, then quote its name, e.g. labels: { "label-name": "value" }.
   */
  labels?: { [name: string]: string };
}

const IBigQueryOptionsProperties = () =>
  strictKeysOf<IBigQueryOptions>()(["partitionBy", "clusterBy", "updatePartitionFilter", "labels"]);

/**
 * Options for creating tables within Presto projects.
 */
export interface IPrestoOptions {
  /**
   * The key with which to partition the table. Typically the name of a timestamp or date column.
   *
   * For more information, read the partitioning documentation for the Presto connection in use.
   */
  partitionBy?: string[];
}

const IPrestoOptionsProperties = () => strictKeysOf<IPrestoOptions>()(["partitionBy"]);

/**
 * Options for creating assertions as part of a dataset definition.
 */
export interface ITableAssertions {
  /**
   * Column(s) which constitute the dataset's unique key index.
   *
   * If set, the resulting assertion will fail if there is more than one row in the dataset with the same values for all of these column(s).
   */
  uniqueKey?: string | string[];

  /**
   * Combinations of column(s), each of which should constitute a unique key index for the dataset.
   *
   * If set, the resulting assertion(s) will fail if there is more than one row in the dataset with the same values for all of the column(s)
   * in the unique key(s).
   */
  uniqueKeys?: string[][];

  /**
   * Column(s) which may never be `NULL`.
   *
   * If set, the resulting assertion will fail if any row contains `NULL` values for these column(s).
   */
  nonNull?: string | string[];

  /**
   * General condition(s) which should hold true for all rows in the dataset.
   *
   * If set, the resulting assertion will fail if any row violates any of these condition(s).
   */
  rowConditions?: string[];
}

const ITableAssertionsProperties = () =>
  strictKeysOf<ITableAssertions>()(["uniqueKey", "uniqueKeys", "nonNull", "rowConditions"]);

/**
 * Configuration options for `dataset` actions, including `table`, `view` and `incremental` action types.
 */
export interface ITableConfig
  extends IActionConfig,
    IDependenciesConfig,
    IDocumentableConfig,
    INamedConfig,
    ITargetableConfig {
  /**
   * The type of the dataset. For more information on how this setting works, check out some of the [guides](guides)
   * on publishing different types of datasets with Dataform.
   */
  type?: TableType;

  /**
   * Only allowed when the table type is `incremental`.
   *
   * If set to true, running this action will ignore the full-refresh option.
   * This is useful for tables which are built from transient data, to ensure that historical data is never lost.
   */
  protected?: boolean;

  /**
   * Redshift-specific warehouse options.
   */
  redshift?: IRedshiftOptions;

  /**
   * BigQuery-specific warehouse options.
   */
  bigquery?: IBigQueryOptions;

  /**
   * Snowflake-specific options.
   */
  snowflake?: ISnowflakeOptions;

  /**
   * Azure SQL Data Warehouse-specific options.
   */
  sqldatawarehouse?: ISQLDataWarehouseOptions;

  /**
   * Presto-specific options.
   */
  presto?: IPrestoOptions;

  /**
   * Assertions to be run on the dataset.
   *
   * If configured, relevant assertions will automatically be created and run as a dependency of this dataset.
   */
  assertions?: ITableAssertions;

  /**
   * Unique keys for merge criteria for incremental tables.
   *
   * If configured, records with matching unique key(s) will be updated, rather than new rows being inserted.
   */
  uniqueKey?: string[];
}

// TODO: This needs to be a method, I'm really not sure why, but it hits a runtime failure otherwise.
export const ITableConfigProperties = () =>
  strictKeysOf<ITableConfig>()([
    "type",
    "disabled",
    "protected",
    "name",
    "redshift",
    "bigquery",
    "snowflake",
    "sqldatawarehouse",
    "presto",
    "tags",
    "uniqueKey",
    "dependencies",
    "hermetic",
    "schema",
    "assertions",
    "database",
    "columns",
    "description"
  ]);

/**
 * Context methods are available when evaluating contextable SQL code, such as
 * within SQLX files, or when using a [Contextable](#Contextable) argument with the JS API.
 */
export interface ITableContext extends ICommonContext {
  /**
   * Shorthand for an `if` condition. Equivalent to `cond ? trueCase : falseCase`.
   * `falseCase` is optional, and defaults to an empty string.
   */
  when: (cond: boolean, trueCase: string, falseCase?: string) => string;

  /**
   * Indicates whether the config indicates the file is dealing with an incremental table.
   */
  incremental: () => boolean;
}

/**
 * @hidden
 */
export class Table {
  public static readonly IGNORED_PROPS: {
    [tableType: string]: Array<keyof dataform.ITable>;
  } = {
    inline: [
      "bigquery",
      "redshift",
      "snowflake",
      "sqlDataWarehouse",
      "presto",
      "preOps",
      "postOps",
      "actionDescriptor",
      "disabled",
      "where"
    ]
  };

  public proto: dataform.ITable = dataform.Table.create({
    type: "view",
    disabled: false,
    tags: []
  });

  // Hold a reference to the Session instance.
  public session: Session;

  // We delay contextification until the final compile step, so hold these here for now.
  public contextableQuery: Contextable<ITableContext, string>;
  private contextableWhere: Contextable<ITableContext, string>;
  private contextablePreOps: Array<Contextable<ITableContext, string | string[]>> = [];
  private contextablePostOps: Array<Contextable<ITableContext, string | string[]>> = [];

  private uniqueKeyAssertions: Assertion[] = [];
  private rowConditionsAssertion: Assertion;

  public config(config: ITableConfig) {
    checkExcessProperties(
      (e: Error) => this.session.compileError(e),
      config,
      ITableConfigProperties(),
      "table config"
    );
    if (config.type) {
      this.type(config.type);
    }
    if (config.dependencies) {
      this.dependencies(config.dependencies);
    }
    if (config.hermetic !== undefined) {
      this.hermetic(config.hermetic);
    }
    if (config.disabled) {
      this.disabled();
    }
    if (config.protected) {
      this.protected();
    }
    if (config.redshift) {
      this.redshift(config.redshift);
    }
    if (config.bigquery) {
      this.bigquery(config.bigquery);
    }
    if (config.snowflake) {
      this.snowflake(config.snowflake);
    }
    if (config.sqldatawarehouse) {
      this.sqldatawarehouse(config.sqldatawarehouse);
    }
    if (config.presto) {
      this.presto(config.presto);
    }
    if (config.tags) {
      this.tags(config.tags);
    }
    if (config.description) {
      this.description(config.description);
    }
    if (config.columns) {
      this.columns(config.columns);
    }
    if (config.database) {
      this.database(config.database);
    }
    if (config.schema) {
      this.schema(config.schema);
    }
    if (config.assertions) {
      this.assertions(config.assertions);
    }
    if (config.uniqueKey) {
      this.uniqueKey(config.uniqueKey);
    }

    return this;
  }

  public type(type: TableType) {
    this.proto.type = type as string;
    return this;
  }

  public query(query: Contextable<ITableContext, string>) {
    this.contextableQuery = query;
    return this;
  }

  public where(where: Contextable<ITableContext, string>) {
    this.contextableWhere = where;
    return this;
  }

  public preOps(pres: Contextable<ITableContext, string | string[]>) {
    this.contextablePreOps.push(pres);
    return this;
  }

  public postOps(posts: Contextable<ITableContext, string | string[]>) {
    this.contextablePostOps.push(posts);
    return this;
  }

  public disabled() {
    this.proto.disabled = true;
    this.uniqueKeyAssertions.forEach(assertion => assertion.disabled());
    this.rowConditionsAssertion?.disabled();
    return this;
  }

  public protected() {
    this.proto.protected = true;
    return this;
  }

  public uniqueKey(uniqueKey: string[]) {
    this.proto.uniqueKey = uniqueKey;
  }

  public snowflake(snowflake: dataform.ISnowflakeOptions) {
    checkExcessProperties(
      (e: Error) => this.session.compileError(e),
      snowflake,
      ISnowflakeOptionsProperties(),
      "snowflake config"
    );
    this.proto.snowflake = dataform.SnowflakeOptions.create(snowflake);
    return this;
  }

  public sqldatawarehouse(sqlDataWarehouse: dataform.ISQLDataWarehouseOptions) {
    checkExcessProperties(
      (e: Error) => this.session.compileError(e),
      sqlDataWarehouse,
      ISQLDataWarehouseOptionsProperties(),
      "sqldatawarehouse config"
    );
    this.proto.sqlDataWarehouse = dataform.SQLDataWarehouseOptions.create(sqlDataWarehouse);
    return this;
  }

  public redshift(redshift: dataform.IRedshiftOptions) {
    checkExcessProperties(
      (e: Error) => this.session.compileError(e),
      redshift,
      IRedshiftOptionsProperties(),
      "redshift config"
    );
    this.proto.redshift = dataform.RedshiftOptions.create(redshift);
    return this;
  }

  public bigquery(bigquery: dataform.IBigQueryOptions) {
    checkExcessProperties(
      (e: Error) => this.session.compileError(e),
      bigquery,
      IBigQueryOptionsProperties(),
      "bigquery config"
    );
    this.proto.bigquery = dataform.BigQueryOptions.create(bigquery);
    if (!!bigquery.labels) {
      if (!this.proto.actionDescriptor) {
        this.proto.actionDescriptor = {};
      }
      this.proto.actionDescriptor.bigqueryLabels = bigquery.labels;
    }
    return this;
  }

  public presto(presto: dataform.IPrestoOptions) {
    checkExcessProperties(
      (e: Error) => this.session.compileError(e),
      presto,
      IPrestoOptionsProperties(),
      "presto config"
    );
    this.proto.presto = dataform.PrestoOptions.create(presto);
    return this;
  }

  public dependencies(value: Resolvable | Resolvable[]) {
    const newDependencies = Array.isArray(value) ? value : [value];
    newDependencies.forEach(resolvable => {
      this.proto.dependencyTargets.push(resolvableAsTarget(resolvable));
    });

    return this;
  }

  public hermetic(hermetic: boolean) {
    this.proto.hermeticity = hermetic
      ? dataform.ActionHermeticity.HERMETIC
      : dataform.ActionHermeticity.NON_HERMETIC;
  }

  public tags(value: string | string[]) {
    const newTags = typeof value === "string" ? [value] : value;
    newTags.forEach(t => {
      this.proto.tags.push(t);
    });
    return this;
  }

  public description(description: string) {
    if (!this.proto.actionDescriptor) {
      this.proto.actionDescriptor = {};
    }
    this.proto.actionDescriptor.description = description;
    return this;
  }

  public columns(columns: IColumnsDescriptor) {
    if (!this.proto.actionDescriptor) {
      this.proto.actionDescriptor = {};
    }
    this.proto.actionDescriptor.columns = ColumnDescriptors.mapToColumnProtoArray(
      columns,
      (e: Error) => this.session.compileError(e)
    );
    return this;
  }

  public database(database: string) {
    setNameAndTarget(
      this.session,
      this.proto,
      this.proto.target.name,
      this.proto.target.schema,
      database
    );
    return this;
  }

  public schema(schema: string) {
    setNameAndTarget(
      this.session,
      this.proto,
      this.proto.target.name,
      schema,
      this.proto.target.database
    );
    return this;
  }

  public assertions(assertions: ITableAssertions) {
    checkExcessProperties(
      (e: Error) => this.session.compileError(e),
      assertions,
      ITableAssertionsProperties(),
      "assertions config"
    );
    if (!!assertions.uniqueKey && !!assertions.uniqueKeys) {
      this.session.compileError(
        new Error("Specify at most one of 'assertions.uniqueKey' and 'assertions.uniqueKeys'.")
      );
    }
    let uniqueKeys = assertions.uniqueKeys;
    if (!!assertions.uniqueKey) {
      uniqueKeys =
        typeof assertions.uniqueKey === "string"
          ? [[assertions.uniqueKey]]
          : [assertions.uniqueKey];
    }
    if (uniqueKeys) {
      uniqueKeys.forEach((uniqueKey, index) => {
        const uniqueKeyAssertion = this.session.assert(
          `${this.proto.target.schema}_${this.proto.target.name}_assertions_uniqueKey_${index}`,
          ctx => this.session.adapter().indexAssertion(ctx.ref(this.proto.target), uniqueKey)
        );
        uniqueKeyAssertion.proto.parentAction = this.proto.target;
        if (this.proto.disabled) {
          uniqueKeyAssertion.disabled();
        }
        this.uniqueKeyAssertions.push(uniqueKeyAssertion);
      });
    }
    const mergedRowConditions = assertions.rowConditions || [];
    if (!!assertions.nonNull) {
      const nonNullCols =
        typeof assertions.nonNull === "string" ? [assertions.nonNull] : assertions.nonNull;
      nonNullCols.forEach(nonNullCol => mergedRowConditions.push(`${nonNullCol} IS NOT NULL`));
    }
    if (!!mergedRowConditions && mergedRowConditions.length > 0) {
      this.rowConditionsAssertion = this.session.assert(
        `${this.proto.target.schema}_${this.proto.target.name}_assertions_rowConditions`,
        ctx =>
          this.session
            .adapter()
            .rowConditionsAssertion(ctx.ref(this.proto.target), mergedRowConditions)
      );
      this.rowConditionsAssertion.proto.parentAction = this.proto.target;
      if (this.proto.disabled) {
        this.rowConditionsAssertion.disabled();
      }
    }
    return this;
  }

  public compile() {
    const context = new TableContext(this);
    const incrementalContext = new TableContext(this, true);

    this.proto.query = context.apply(this.contextableQuery);

    if (this.proto.type === "incremental") {
      this.proto.incrementalQuery = incrementalContext.apply(this.contextableQuery);

      this.proto.incrementalPreOps = this.contextifyOps(this.contextablePreOps, incrementalContext);
      this.proto.incrementalPostOps = this.contextifyOps(
        this.contextablePostOps,
        incrementalContext
      );
    }

    if (this.contextableWhere) {
      this.proto.where = context.apply(this.contextableWhere);
    }

    this.proto.preOps = this.contextifyOps(this.contextablePreOps, context);
    this.proto.postOps = this.contextifyOps(this.contextablePostOps, context);

    validateQueryString(this.session, this.proto.query);
    this.proto.preOps.forEach(preOp => validateQueryString(this.session, preOp));
    this.proto.postOps.forEach(postOp => validateQueryString(this.session, postOp));

    return this.proto;
  }

  private contextifyOps(
    contextableOps: Array<Contextable<ITableContext, string | string[]>>,
    currentContext: TableContext
  ) {
    let protoOps: string[] = [];
    contextableOps.forEach(contextableOp => {
      const appliedOps = currentContext.apply(contextableOp);
      protoOps = protoOps.concat(typeof appliedOps === "string" ? [appliedOps] : appliedOps);
    });
    return protoOps;
  }
}

/**
 * @hidden
 */
export class TableContext implements ITableContext {
  constructor(private table: Table, private isIncremental = false) {}

  public config(config: ITableConfig) {
    this.table.config(config);
    return "";
  }

  public self(): string {
    return this.resolve(this.table.proto.target);
  }

  public name(): string {
    return this.table.proto.target.name;
  }

  public ref(ref: Resolvable | string[], ...rest: string[]): string {
    ref = toResolvable(ref, rest);
    if (!resolvableAsTarget(ref)) {
      const message = `Action name is not specified`;
      this.table.session.compileError(new Error(message));
      return "";
    }
    this.table.dependencies(ref);
    return this.resolve(ref);
  }

  public resolve(ref: Resolvable | string[], ...rest: string[]) {
    return this.table.session.resolve(toResolvable(ref, rest));
  }

  public type(type: TableType) {
    this.table.type(type);
    return "";
  }

  public where(where: Contextable<ITableContext, string>) {
    this.table.where(where);
    return "";
  }

  public when(cond: boolean, trueCase: string, falseCase: string = "") {
    return cond ? trueCase : falseCase;
  }

  public incremental() {
    return !!this.isIncremental;
  }

  public preOps(statement: Contextable<ITableContext, string | string[]>) {
    this.table.preOps(statement);
    return "";
  }

  public postOps(statement: Contextable<ITableContext, string | string[]>) {
    this.table.postOps(statement);
    return "";
  }

  public disabled() {
    this.table.disabled();
    return "";
  }

  public redshift(redshift: dataform.IRedshiftOptions) {
    this.table.redshift(redshift);
    return "";
  }

  public bigquery(bigquery: dataform.IBigQueryOptions) {
    this.table.bigquery(bigquery);
    return "";
  }

  public presto(presto: dataform.IPrestoOptions) {
    this.table.presto(presto);
    return "";
  }

  public dependencies(res: Resolvable) {
    this.table.dependencies(res);
    return "";
  }

  public apply<T>(value: Contextable<ITableContext, T>): T {
    if (typeof value === "function") {
      return (value as any)(this);
    } else {
      return value;
    }
  }

  public tags(tags: string[]) {
    this.table.tags(tags);
    return "";
  }
}

File: pycore/table.py
from __future__ import annotations

from typing import Literal, Optional, List, Dict
from dataclasses import dataclass
from protos.core_pb2 import (
    Target,
    Table as TableProto,
    TableType,
    ProjectConfig,
    ColumnDescriptor,
)
from common import (
    ActionConfig,
    action_target,
    target_to_target_representation,
    efficient_replace_string,
)
from pathlib import Path
from assertion import Assertion
import adapter
from google.protobuf import json_format


# TODO: These are all snake case, but the current open source uses camel case. We shouldn't
# necessarily convert them to camel case however.
@dataclass
class TableAssertions:
    """
    A data class that represents table assertions.

    Args:
        unique_key: The column(s) which constitute the dataset's unique key index.
        unique_keys: Combinations of column(s), each of which should constitute a unique key index for the dataset.
        non_null: The column(s) which may never be `NULL`.
        row_conditions: General condition(s) which should hold true for all rows in the dataset.
    """

    unique_key: Optional[List[str]] = None  # TODO: Can be just str type too.
    unique_keys: Optional[List[List[str]]] = None
    non_null: Optional[List[str]] = None  # TODO: Can be just str type too.
    row_conditions: Optional[List[str]] = None


@dataclass
class TableConfig(ActionConfig):
    """
    A data class that represents table configuration options.

    Args:
        type: The type of the dataset. For more information on how this setting works, check out some of the [guides](guides) on publishing different types of datasets with Dataform.
        protected: Optional[bool]: Only allowed when the table type is `incremental`.
            If set to true, running this action will ignore the full-refresh option.
            This is useful for tables which are built from transient data, to ensure that historical data is never lost.
        assertions: Optional[TableAssertions]: Assertions to be run on the dataset.
            If configured, relevant assertions will automatically be created and run as a dependency of this dataset.
        unique_key: Optional[List[str]]: Unique keys for merge criteria for incremental tables.
            If configured, records with matching unique key(s) will be updated, rather than new rows being inserted.
        materialized: Only valid when the table type is `view`.
            If set to true, will make the view materialized.
            For more information, read the [BigQuery materialized view docs](https://cloud.google.com/bigquery/docs/materialized-views-intro).

        partition_by: The key with which to partition the table. Typically the name of a timestamp or date column.
        cluster_by: The keys by which to cluster partitions by.
        update_partition_filter: SQL based filter for when incremental updates are applied.
        labels: Key-value pairs for [BigQuery labels](https://cloud.google.com/bigquery/docs/labels-intro).
        partition_expiration_days: This setting specifies how long BigQuery keeps the data in each partition.
        require_partition_filter: When you create a partitioned table, you can require that all queries on the table must include a predicate filter (a WHERE clause) that filters on the partitioning column.

        additional_options: Key-value pairs for options [table](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#table_option_list), [view](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#view_option_list), [materialized view](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#materialized_view_option_list).
    """

    type: Optional[TableType] = None
    protected: Optional[bool] = None
    assertions: Optional[TableAssertions] = None
    unique_key: Optional[List[str]] = None
    materialized: Optional[bool] = None

    # These were originally BigQuery specific options.
    partition_by: Optional[str] = None
    cluster_by: Optional[List[str]] = None
    update_partition_filter: Optional[str] = None
    labels: Optional[Dict[str, str]] = None
    partition_expiration_days: Optional[int] = None
    require_partition_filter: Optional[bool] = None

    # Additional options can be used by hookable adapters.
    additional_options: Optional[Dict[str, str]] = None


class Table:
    def __init__(
        self,
        project_config: ProjectConfig,
        path: Path,
        session,  # TODO: Can't import this type because cirular dependency; maybe spread session?
        table_type: TableType,
        table_config_as_map: Dict,
    ):
        self._project_config = project_config
        self._path = path
        self._session = session
        self._table_type = table_type
        self._table_config = TableConfig(**table_config_as_map)

        self._proto = TableProto()

        # Canonical target is based off of the file structure, and is guaranteed to be unique.
        self._proto.canonical_target.CopyFrom(action_target(project_config, path.stem))

        # Target is the final resolved target, and can be overridden by the table config.
        self._proto.target.CopyFrom(
            action_target(
                project_config,
                path.stem,
                self._table_config.database,
                self._table_config.schema,
                self._table_config.name,
            )
        )

        self._populate_proto_fields()
        self._add_assertions()
        self._populate_bigquery_proto_fields()

    def _populate_proto_fields(self):
        if TableType == TableType.UNKNOWN_TYPE:
            raise Exception(f"Unknown table type: {self._table_config.type}")
        else:
            self._proto.enum_type = self._table_type
        self._proto.disabled = bool(self._table_config.disabled)
        if self._table_config.protected:
            self._proto.protected = self._table_config.protected
        if self._table_config.tags:
            self._proto.tags.extend(self._table_config.tags)
        if self._table_config.description:
            self._proto.action_descriptor.description = self._table_config.description
        if self._table_config.columns:
            self._proto.action_descriptor.columns.extend(
                [
                    json_format.ParseDict(column, ColumnDescriptor())
                    for column in self._table_config.columns
                ]
            )
        if self._table_config.unique_key:
            self._proto.unique_key = self._table_config.uniqueKey
        if self._table_config.materialized:
            self._proto.materialized = self._table_config.materialized
        if self._table_config.dependencies:
            self._proto.dependency_targets.extend(
                [
                    self._session._resolve_target_representation(target_representation)
                    for target_representation in self._table_config.dependencies
                ]
            )
        self._proto.file_name = str(self._path)

    def _add_assertions(self):
        if self._table_config.assertions:
            assertions = TableAssertions(**self._table_config.assertions)
            if assertions.unique_key and assertions.unique_keys:
                raise Exception(
                    "Specify at most one of 'assertions.unique_key' and 'assertions.unique_keys'."
                )
            if assertions.unique_key:
                assertions.unique_keys = [assertions.unique_key]
            assertion_config_to_share = {
                "tags": self._proto.tags,
                "disabled": self._proto.disabled,
                "dependencies": [self.target_representation()],
            }
            for unique_key in assertions.unique_keys:
                self._session._add_action(
                    Assertion(
                        self._project_config,
                        self._path.parent
                        # TODO: Confirm this is an improvement: it differs from previous
                        # implementation, but facilitates multiple unique key assertions in the same
                        # file.
                        / f"{self._proto.target.name}_assertions_unique_key_{'_'.join(unique_key)}",
                        self._session,
                        assertion_config_to_share,
                    ).sql(
                        adapter.index_assertion(
                            self.target_representation(), unique_key
                        )
                    )
                )

            merged_row_conditions = assertions.row_conditions or []
            if assertions.non_null:
                merged_row_conditions.extend(
                    [f"{col} IS NOT NULL" for col in assertions.non_null]
                )
            if len(merged_row_conditions) > 0:
                self._session._add_action(
                    Assertion(
                        self._project_config,
                        self._path.parent
                        / f"{self._path.stem}_assertions_rowConditions",
                        self._session,
                        assertion_config_to_share,
                    ).sql(
                        adapter.row_conditions_assertion(
                            self.target_representation(), merged_row_conditions
                        )
                    )
                )

    def _populate_bigquery_proto_fields(self):
        """These are options that were previously BigQuery specific."""
        if self._table_config.partition_by:
            self._proto.bigquery.partition_by = self._table_config.partition_by
        if self._table_config.cluster_by:
            self._proto.bigquery.cluster_by = self._table_config.cluster_by
        if self._table_config.update_partition_filter:
            self._proto.bigquery.update_partition_filter = (
                self._table_config.update_partition_filter
            )
        if self._table_config.labels:
            self._proto.bigquery.labels = self._table_config.labels
        if self._table_config.partition_expiration_days:
            self._proto.bigquery.partition_expiration_days = (
                self._table_config.partition_expiration_days
            )
        if self._table_config.require_partition_filter:
            self._proto.bigquery.require_partition_filter = (
                self._table_config.require_partition_filter
            )
        if self._table_config.additional_options:
            # TODO: Expand dynamically.
            self._proto.bigquery.additional_options = (
                self._table_config.additional_options
            )
        if self._table_config.labels:
            self._proto.action_descriptor.bigquery_labels = self._table_config.labels

    def sql(self, sql: str) -> "Table":
        if self._proto.query:
            raise Exception("SQL is already defined")
        self._proto.query = sql
        if self._proto.enum_type == TableType.INCREMENTAL:
            self._proto.incremental_query = sql
        # TODO: Validate query strings (remove semi-colon at end), or no longer necessary?
        return self

    def pre_operations(self, sqls: List[str]) -> "Table":
        self._proto.pre_ops.extend(sqls)
        if self._proto.enum_type == TableType.INCREMENTAL:
            self._proto.incremental_pre_ops.extend(sqls)
        return self

    def post_operations(self, sqls: List[str]) -> "Table":
        self._proto.post_ops.extend(sqls)
        if self._proto.enum_type == TableType.INCREMENTAL:
            self._proto.incremental_post_ops.extend(sqls)
        return self

    def load_sql_file(self, sql_file_path_as_str: str) -> "Table":
        self.sql(self._session._load_sql_file(sql_file_path_as_str))
        return self

    def _add_dependency(self, target: Target):
        self._proto.dependency_targets.append(target)

    def target_representation(self):
        return target_to_target_representation(self._proto.target)

    def canonical_target_representation(self):
        return target_to_target_representation(self._proto.canonical_target)

    def clean_refs(self, refs_to_replace: Dict[str, str]):
        if not refs_to_replace:
            return
        self._proto.query = efficient_replace_string(refs_to_replace, self._proto.query)
        # TODO: Clean pre-ops and post-ops.
